### DISCUSSION TITLE
Modding refactor Feedback and user accessibility

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#8682] [Unstable feedback]
**1 The very long modding list**
The mod list gets very, very long, with every single item being listed as its own mod. Not just the downloaded ones, but every sub, assembly, wreck, outpost, beacon... If you are a somewhat active sub builder, you easily end up with 100+ items on said list. 

You end up with a lot of noise caused by items that are **never going to leave the sub editor**:
![image](https://user-images.githubusercontent.com/73229309/161081916-f707a4ec-a9e4-4a79-8171-6aa0de784886.png)

_It also affects the sub editor content package list! Assemblies aren't "content packages"!_
![image](https://user-images.githubusercontent.com/73229309/161061497-67f1b390-fd0b-4c7f-a5f1-517b8aaa661a.png)

It is hard to visually parse what's on the list, as all "mods" are equally represented by a single line. 
**In practice however there's a hierarchy of mods based on how they impact the server you're hosting:** 

![image](https://user-images.githubusercontent.com/73229309/161053392-537641ea-d8a8-44a1-8c2e-4321da2bea94.png)

On both ends of the spectrum, 90% of **Assemblies** will only be used locally, a few may be uploaded, and they don't show up when hosting a server. Assemblies are low-priority yet take the same space as... **Total conversions**, like ITA or Thalassophobia.

Mods need to be sorted by "hierarchy" and types in the list, and/or need icons to parse the contents of mods. Total conversions will typically have custom subs+enemies+items+outposts files. The list needs to reflect this and tell you "this is a big important mod" and "this is just a local assembly"

---

**2 Editing and Hosting are (more than before) in conflict**
Now, you need to enable your own local files in the mods to see them in the editor. But it also means that, when you're hosting, these files are enabled. You cannot freely edit something that you do not mean to host, and vice versa, **without having to constantly enable and disable packages.** 

The refactor made it easier to host certain types of custom content, like wrecks, which are now automatically spawned in the level. But if you are working on a wreck, or a beacon, or any WIP project, you'll need to disable it to not host it. If you are an active host and editor at the same time, it is tedious.

**The best of both worlds would be to have control over what appears in the editor and what mods are enabled on the server separately.** Submarines already do that by having the "Hide in menus", which makes it practically impossible to use this submarine file in a mutliplayer server. Editing and hosting shouldn't be so much in conflict.

---

**3 Not all file types should be enabled and immediately "hostable" by default.**
Wrecks/Beacons/Outposts... are files you'll be _constantly_ working on, most likely released in bundled packages. **A single line in the .xml should make the difference between a file that is meant to be edited, and a file that is playable or can be hosted as a working mod on a server.**
It's nice that you can easily host custom wrecks, but not if it comes at the cost of fighting the UI

---

**4 Your average Host has no clue what they're making others download**
Chances are, hosts will keep forgetting that they have left some mods enabled.
If you are hosting, the game does not tell you what mods are activated at all, without manually going in the settings menu and its bloated 100+ items mod list (see 1.), with no sorting or visual clues of what actually affects the server and what doesn't (submarines, assemblies)

You need to make it clearer what content packages, not submarines, not assemblies, not random WIP wrecks (see 3.) are affecting the server.

_Also, if you leave and rejoin a server, you apparently download all the mods twice, as the game doesn't check if your versions match but this has to be confirmed._

---

**5 Random content package save prompts**
The game tries to help by prompting you to save items in a similar content package. For example, wrecks into a "wreck" content package/folder.
But this stops working if you have multiple of those. If you have multiple wreck files, the  game can't decide which folder it will use.
![image](https://user-images.githubusercontent.com/73229309/161060874-851a5124-75fc-4e09-9317-d6d85b44a72b.png)
![image](https://user-images.githubusercontent.com/73229309/161061029-33d6bc77-f097-4d96-bf08-348b6726f3a6.png)

The prompt should merely *remind you* to check a content package from the list
![image](https://user-images.githubusercontent.com/73229309/161061184-7c7490f2-0ab9-4ec6-a22d-e6dfacde1282.png)

**Comment by cshingle** [comment]
It would be great if the UI had an intuitive way of adding multiple subs to the same mod.

**Comment by evilfactory** [comment]
The content package UI should have two separate tabs for multiplayer compatible mods and non multiplayer compatible ones, to better distinguish between which mods players will need to download when joining the server, and which ones they don't.

**Comment by rav2n** [comment]
For subbuilders, you need to keep access to all custom submarines from the submarine editor. Without the need to enable or disable them from the main menu every time, like a full-fledged mod (as indicated above, otherwise what's the point of dividing mods into two categories). It can also become an annoying factor that when creating a copy of the submarine, a separate mod is created.

A good solution would be to keep the old functionality of the "Submarines" folder (Barotrauma\Submarines), but only for user works. With the loading of these files only at the time of use inside the editor or for the game.
Vanilla submarines will take root perfectly in the new folder (Barotrauma\Content\Submarines).

**Reply by ubertpendragon** [reply]
Absolutely. 

Editing needs to be accessible at all time for Local Mods, whether or not the mod is activated in multiplayer.

**Comment by cshingle** [comment]
It would also be nice if there is a way for the server to have clients get Steam Workshop mods from the Steam Workshop like the current stable. Requiring clients to download numerous large mods from the server every time they restart the game will be a big issue. Mods that add new creatures tend to be fairly large.

Edit:
Looks like this will not be an issue
 > c3679fceca: Added a button to first subscribe to whatever Workshop items are currently not installed before attempting to download from the server directly

**Comment by Ek-Videogames** [comment]
**Literally everything** needs search bar functionality - bonus points for adding sort alphabetically & sort by date.
 
 Also de-convolute submarine file management needs to be a priority.
